app/src/main/java/org/dhis2/utils/category/CategoryDialogPresenter.kt

```diff
@@ -0,0 +1,192 @@
+                .subscribe(
+                    { view.setTitle(it.displayName() ?: "-") },
+                    { Timber.e(it) }
+                )
+        )
+        disposable.add(
+            view.searchSource()
+                .debounce(500, TimeUnit.MILLISECONDS, schedulerProvider.io())
+                .map { textToSearch ->
+                    mapCategoryOptionToLivePageList(textToSearch)
+                }
+                .subscribeOn(schedulerProvider.io())
+                .observeOn(schedulerProvider.ui())
+                .subscribe(
+                    { view.setLiveData(it) },
+                    { Timber.e(it) }
+                )
+        )
+    }
+
+    private fun mapCategoryOptionToLivePageList(
+        textToSearch: String
+    ): LiveData<PagedList<CategoryDialogItem>> {
+        var catOptionRepository = d2.categoryModule().categoryOptions()
+            .byCategoryUid(uid)
+
+        if (textToSearch.isNotEmpty()) {
+            catOptionRepository =
+                catOptionRepository.byDisplayName().like("%$textToSearch%")
+        }
+
+        if (withAccessControl) {
+            catOptionRepository =
+                catOptionRepository.byAccessDataWrite().isTrue
+        }
+
+        val dataSource =
+            catOptionRepository.orderByDisplayName(RepositoryScope.OrderByDirection.ASC)
+                .dataSource
+                .mapByPage { options -> filterByDate(options) }
+                .map { catOption -> catOptMapper.map(catOption) }
+
+        return LivePagedListBuilder(
+            object : DataSource.Factory<CategoryOption, CategoryDialogItem>() {
+                override fun create(): DataSource<CategoryOption, CategoryDialogItem> {
+                    return dataSource
+                }
+            },
+            20
+        ).build()
+    }
+
+    private fun mapCatOptComboToLivePageList(
+        textToSearch: String
+    ): LiveData<PagedList<CategoryDialogItem>> {
+        var catOptComboRepository = d2.categoryModule().categoryOptionCombos()
+            .byCategoryComboUid().eq(uid)
+
+        if (textToSearch.isNotEmpty()) {
+            catOptComboRepository =
+                catOptComboRepository.byDisplayName().like("%$textToSearch%")
+        }
+
+        val dataSource =
+            catOptComboRepository
+                .orderByDisplayName(RepositoryScope.OrderByDirection.ASC)
+                .dataSource
+                .map { catOption -> catOptCombMapper.map(catOption) }
+
+        return createDataSource(dataSource)
+    }
+
+    private fun filterByDate(options: MutableList<CategoryOption>): MutableList<CategoryOption>? {
+        val iterator = options.iterator()
+        while (iterator.hasNext()) {
+            val option = iterator.next()
+            if (date != null &&
+                (isBeforeDate(option) || isAfterDate(option))
+            ) {
+                iterator.remove()
+            }
+        }
+        return options
```

Comment:
why not using options.filterNot{ isBeforeDate(option) || isAfterDate(option)) } ???

Reply:
The filter method creates a new list, and this is not supported by the paged library without it crashing (similar to what happened in the search screen)